### PR TITLE
feat(feeds): Commands can now have implicitly use parameters from feed in VM

### PR DESCRIPTION
**Pull request by dr1rrb**
## Feature
Commands can now have implicitly use parameters from feed in VM

## What is the current behavior?
You need to await parameters / feeds that are declared on the VM in your command

## What is the new behavior?
If a parameter name match the name and type of a Feed in your VM, the parameter is automatically full-filed.

## PR Checklist
- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

### CODE REVIEW

File: doc/Overview/Reactive/in-apps.md [hunk 1]
```diff
@@ -103,8 +103,7 @@ so you can enhance the UX of the pagination, like display a loading indicator wh
 
 
 ## Commands
-The generated bindable counterpart of a class will automatically re-expose public methods that have 0 or 1 parameter and an optional `CancellationToken` as `ICommand`.
-The parameter will be fulfilled using the `CommandParameter` property.
+The generated bindable counterpart of a class will automatically re-expose as `ICommand` public methods that are compatible (cf. below).
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:106]: Compatible with what ?

File: doc/Overview/Reactive/in-apps.md [hunk 2]
```diff
@@ -120,3 +119,53 @@ This will be exposed into an `ICommand` that can be data-bound to the `Command`
 ```xml
 <Button Command="{Binding Share}" Content="Share" />
 ```
+
+By default, if the method has a parameter `T myValue` and there is a property `Feed<T> MyValue` on the same class (matching type and name), 
+that parameter will be automatically full-filed from that feed.
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:124]:
```suggestion

By default, if the method has a parameter `T myValue` and there is a property `Feed<T> MyValue` on the same class (matching type and name), that parameter will automatically be filled from that feed.
```

File: doc/Overview/Reactive/in-apps.md [hunk 3]
```diff
@@ -120,3 +119,53 @@ This will be exposed into an `ICommand` that can be data-bound to the `Command`
...
+<!-- Without CommandParameter -->
+<!-- 'message' arg in the 'Share' method will be the current value of the Message _feed_ -->
+<Button Command="{Binding Share}" Content="Share" />
+
+<!-- With CommandParameter -->
+<!-- 'message' arg in the 'Share' method will be "hello world" -->
+<Button Command="{Binding Share}" CommandParameter="hello world" Content="Share" />
+```
+
+You can even mix "feed parameters" with the "view parameter" (i.e. the value of the `CommandParameter`):
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:149]:
```suggestion
You can also use both "feed parameters" and "view parameter" (i.e. the value of the `CommandParameter`):
```

File: doc/Overview/Reactive/in-apps.md [hunk 4]
```diff
@@ -120,3 +119,53 @@ This will be exposed into an `ICommand` that can be data-bound to the `Command`
...
+
+```csharp
+public IFeed<MyEntity> Entity { get; }
+
+public async ValueTask Share(MyEntity myEntity, string origin)
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:154]: Why isn't the parameter `entity`? How does the matching work?

File: doc/Overview/Reactive/in-apps.md [hunk 5]
```diff
@@ -120,3 +119,53 @@ This will be exposed into an `ICommand` that can be data-bound to the `Command`
...
+public async ValueTask Share(MyEntity myEntity, string origin) 
+{
+}
+```
+
+```xml
+<Button Command="{Binding Share}" CommandParameter="command_bar" Content="Share" />
+```
+
+General rules for you method to be re-exposed as commands are:
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:163]:
```suggestion
General rules for methods to be re-exposed as commands are:
```

File: doc/Overview/Reactive/in-apps.md [hunk 6]
```diff
@@ -120,3 +119,53 @@ This will be exposed into an `ICommand` that can be data-bound to the `Command`
...
+* At most one paramater that cannot be resolved from a _Feed_  property in your VM (a.k.a the `CommandParameter`);
+* At most one `CancellationToken`;
+* Method can be sync, or async
+
+> [!NOTE]
+> The automatic parameters resolution be configured using attributes:
+> - `[ImplicitFeedCommandParameters]` on assembly or class to enable or disable the implicit parameters resolution.
+> - `[FeedParameter("MyEntity")]` to explicit the property to use.
```
Comment (jeromelaban) [inline doc/Overview/Reactive/in-apps.md:171]: This probably needs an explicit example, this is rather important :)

File: src/Uno.Extensions.Reactive.Generator/Rules.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+{
+	// references:
+	//	Categories: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/categories
+
+	// General usage [0000-0999]
+
+	// Bindings [1000-1999]
+
+	// Commands [2000-2999]
+	public static class FEED2001
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.Generator/Rules.cs:20]: Those should be in the documentation, also, having a call to action in the message would be useful.

File: src/Uno.Extensions.Reactive.Generator/Rules.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+					? Location.Create(syntax.SyntaxTree, syntax.Span)
+					: Location.None,
+				parameter.Name,
+				@class.Name,
+				method.Name);
+	}
+
+	public static class FEED2002
+	{
+		private const string message = "The property '{0}' resolved on the class '{1}' cannot be used as command parameter for '{2}' as it's not a Feed";
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.Generator/Rules.cs:48]:
```suggestion
		private const string message = "The property '{0}' resolved on the class '{1}' is not a Feed. It cannot be used as command parameter for '{2}'.";
```

File: src/Uno.Extensions.Reactive.Generator/Rules.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+		public static readonly DiagnosticDescriptor Descriptor = new DiagnosticDescriptor(
+			nameof(FEED2002),
+			"Invalid property type",
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.Generator/Rules.cs:52]: Invalid what way ? What can be done?

File: src/Uno.Extensions.Reactive.Generator/Utils/StringExtensions.cs [hunk 10]
```diff
@@ -109,4 +110,7 @@ public static string Replace(this string text, char[] chars, char replacement)
 
 		return result.ToString();
 	}
+
+	public static string RemoveEmptyLines(this string text)
+		=> Regex.Replace(text, "^\\s*$", string.Empty, RegexOptions.Multiline);
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.Generator/Utils/StringExtensions.cs:115]: Would benefit from being a static compiled member.

File: src/Uno.Extensions.Reactive.UI/Config/ImplicitFeedCommandParametersAttribute.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Windows.Input;
+
+namespace Uno.Extensions.Reactive.UI.Config;
+
+/// <summary>
+/// Defines for generated commands if parameters should be automatically full-filled based on public feeds properties exposed by the class.
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.UI/Config/ImplicitFeedCommandParametersAttribute.cs:9]:
```suggestion
/// Defines for generated commands whether parameters should be automatically populated based on public feeds properties exposed by the class.
```

File: src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Windows.Input;
+using Uno.Extensions.Reactive.UI.Config;
+
+namespace Uno.Extensions.Reactive.UI.Presentation.Commands;
+
+/// <summary>
+/// Flags a parameter of a method this is being exposed as <see cref="ICommand"/> (cf. <see cref="CommandAttribute"/>),
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs:10]:
```suggestion
/// Flags a parameter of a method which is being exposed as <see cref="ICommand"/> (cf. <see cref="CommandAttribute"/>),
```

File: src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+/// to be automatically full-filled from a feed declared on the same class.
+/// </summary>
+/// <remarks>
+/// This attribute is optional if you enabled <see cref="ImplicitFeedCommandParametersAttribute"/> in your project
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs:14]:
```suggestion
/// This attribute is optional if <see cref="ImplicitFeedCommandParametersAttribute"/> is enabled in the project
```

File: src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+/// and you have a public feed that matches type and name of the parameter.
+/// </remarks>
+[AttributeUsage(AttributeTargets.Parameter)]
+public class FeedParameterAttribute : Attribute
+{
+	/// <summary>
+	/// Flags a parameter of a method this is being exposed as <see cref="ICommand"/> (cf. <see cref="CommandAttribute"/>),
+	/// to be automatically full-filled from a feed declared on the same class.
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs:22]:
```suggestion
	/// to be automatically populated from a feed declared on the same class.
```

File: src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+	/// </summary>
+	/// <param name="property">The name of the property that should be used as input for the parameter.</param>
```
Comment (jeromelaban) [inline src/Uno.Extensions.Reactive.UI/Presentation/Commands/FeedParameterAttribute.cs:24]:
```suggestion
	/// <param name="property">The name of the property which should be used as an input for the parameter.</param>
```

**Review by jeromelaban — APPROVED** [review]

File: doc/Overview/Reactive/in-apps.md (hunk 4 above)
Comment (dr1rrb) [reply doc/Overview/Reactive/in-apps.md:154]: Humm, indeed the parameter should be `entity`. Matching works on type and name (explained above)

```suggestion
public async ValueTask Share(MyEntity entity, string origin) 
```

File: src/Uno.Extensions.Reactive.Generator/Rules.cs (hunk 9 above)
Comment (dr1rrb) [reply src/Uno.Extensions.Reactive.Generator/Rules.cs:52]: This is only the "title", the "message" already gives more details, and I'm adding a direct link to the doc

File: src/Uno.Extensions.Reactive.Generator/Utils/StringExtensions.cs (hunk 10 above)
Comment (dr1rrb) [reply src/Uno.Extensions.Reactive.Generator/Utils/StringExtensions.cs]: This was not used, removed

File: src/Uno.Extensions.Reactive.Generator/Rules.cs (hunk 7 above)
Comment (dr1rrb) [reply src/Uno.Extensions.Reactive.Generator/Rules.cs:20]: I've added direct link to the doc

File: doc/Overview/Reactive/in-apps.md (hunk 1 above)
Comment (dr1rrb) [reply doc/Overview/Reactive/in-apps.md:106]: Rules are explain "below", but there is no anchor that I can refer to :/

**Pull request merged**